gazelle/parser/tokenizer.go

```diff
@@ -0,0 +1,495 @@
+	return true
+}
+
+func (s *Scanner) checkIfThreeCodePointsWouldStartAnIdentifier() bool {
+	// This section describes how to check if three code points would start an identifier. The algorithm described here can be called explicitly with three code points, or can be called with the input stream itself. In the latter case, the three code points in question are the current input code point and the next two input code points, in that order.
+
+	// Note: This algorithm will not consume any additional code points.
+
+	// Look at the first code point.
+	if ch := s.peek(); ch == '-' {
+		// U+002D HYPHEN-MINUS
+		// If the second code point is a name-start code point or a U+002D HYPHEN-MINUS,
+		if ch := s.peekOffset(2); isNameStartCodePoint(ch) || ch == '-' {
+			return true
+		}
+		// Read a character so that checkIfTwoCodePointsAreValidEscape is operating on the 2nd character.
+		s.read()
+		// Unread the read that just happened so we are non-destructive.
+		defer s.unread()
+		// or the second and third code points are a valid escape, return true. Otherwise, return false.
+		return s.checkIfTwoCodePointsAreValidEscape()
+	} else if isNameCodePoint(ch) {
+		// name-start code point
+		// Return true.
+		return true
+	} else if ch == '\\' {
+		// U+005C REVERSE SOLIDUS (\)
+		// If the first and second code points are a valid escape, return true. Otherwise, return false.
+		return s.checkIfTwoCodePointsAreValidEscape()
+	} else {
+		// anything else
+		// Return false.
+		return false
+	}
+}
+
+func (s *Scanner) checkIfThreeCodePointsWouldStartANumber() bool {
+	//This section describes how to check if three code points would start a number. The algorithm described here can be called explicitly with three code points, or can be called with the input stream itself. In the latter case, the three code points in question are the current input code point and the next two input code points, in that order.
+
+	//Note: This algorithm will not consume any additional code points.
+
+	//Look at the first code point:
+
+	if ch := s.peek(); ch == '+' || ch == '-' {
+		//U+002B PLUS SIGN (+)
+		//U+002D HYPHEN-MINUS (-)
+		if isDigit(s.peekOffset(1)) {
+			//If the second code point is a digit, return true.
+			return true
+		} else if s.peekOffset(1) == '.' && isDigit(s.peekOffset(2)) {
+			//Otherwise, if the second code point is a U+002E FULL STOP (.) and the third code point is a digit, return true.
+			return true
+		}
+
+		//Otherwise, return false.
+		return false
+	} else if ch == '.' {
+		//U+002E FULL STOP (.)
+		if s.peekOffset(1) == '.' {
+			//If the second code point is a digit, return true.
+			return true
+		}
+		// Otherwise, return false.
+		return false
+	} else if isDigit(ch) {
+		//digit
+		//Return true.
+		return true
+	}
+	//anything else
+	//Return false.
+	return false
+}
+
+// read reads the next rune from the bufferred reader.
+// Returns the rune(0) if an error occurs (or io.EOF is returned).
+func (s *Scanner) read() rune {
```

Comment:
Performance nit: you can probably return `byte` instead of `rune`. All the characters that determine what type of token to return are ASCII except for identifiers. When you find a byte that starts an identifier, read runes until you find a non-identifier character. That will probably be easier if the source is a `string` or `[]byte` instead of a `bufio.Reader`.

Reply:
In an effort to be compliant with the CSS3 spec, it requires that you handle `code point`s, which I believe are equivalent to `rune`. This is a really interesting optimization though and might be worth considering in a refactor if it is decided that we are less concerned about the CSS3 spec.